app/state.go

```diff
@@ -1847,3 +1855,207 @@ func setBrokenArtifactFlag(store store.Store, artName string) {
+	}
+}
+
+type fetchControlMapState struct {
+	baseState
+	update       *datastore.UpdateInfo
+	wrappedState UpdateState
+}
+
+func NewFetchControlMapState(wrappedState UpdateState) State {
+	return &fetchControlMapState{
+		baseState: baseState{
+			id: datastore.MenderStateFetchUpdateControl,
+			t:  ToNone,
+		},
+		wrappedState: wrappedState,
+	}
+}
+
+func (c *fetchControlMapState) Handle(ctx *StateContext, controller Controller) (State, bool) {
+
+	log.Debugf("Handle fetchControlMap state")
+
+	// The update control maps are magically updated in here
+	_, err := controller.CheckUpdate()
+
+	if err != nil {
+		if errors.Is(err, client.ErrNoDeploymentAvailable) {
+			return c.wrappedState.HandleError(ctx, controller,
+				NewTransientError(errors.New("The deployment was aborted from the server")))
+		}
+
+		log.Errorf("Update control map check failed: %s, retrying...", err.Error())
+		return NewFetchRetryControlMapState(c.wrappedState), false
+	}
+
+	return NewControlMapState(c.wrappedState), false
+}
+
+type fetchRetryControlMapState struct {
+	waitState
+	wrappedState UpdateState
+	retries      int
+}
+
+func NewFetchRetryControlMapState(wrappedState UpdateState) State {
+	return &fetchRetryControlMapState{
+		waitState: waitState{
+			baseState: baseState{
+				id: datastore.MenderStateFetchRetryUpdateControl,
+				t:  ToNone,
+			},
+		},
+		wrappedState: wrappedState,
+	}
+}
+
+func (f *fetchRetryControlMapState) Handle(ctx *StateContext, c Controller) (State, bool) {
+
+	log.Debugf("Handle fetch update control retry state")
+
+	intvl, err := client.GetExponentialBackoffTime(f.retries, c.GetUpdatePollInterval())
+	if err != nil {
+		return NewUpdateErrorState(
+			NewTransientError(err), f.wrappedState.Update()), false
+	}
+
+	f.retries++
+
+	log.Infof("Wait %v before next update control map fetch/update attempt", intvl)
+	return f.Wait(
+		NewFetchControlMapState(f.wrappedState),
+		f,
+		intvl,
+		ctx.WakeupChan,
+	)
+}
+
+type controlMapPauseState struct {
+	waitState
+	wrappedState UpdateState
+}
+
+func NewControlMapPauseState(wrappedState UpdateState) State {
+	return &controlMapPauseState{
+		waitState: waitState{
+			baseState: baseState{
+				id: datastore.MenderStateUpdateControlPause,
+				t:  ToNone,
+			},
+		},
+		wrappedState: wrappedState,
+	}
+}
+
+// Pause until 1 of two events:
+//
+// 1. a timeout, which means we need to refresh the map
+// 2. an update map expires, which means we need to query the next action
+//
+func (c *controlMapPauseState) Handle(ctx *StateContext, controller Controller) (State, bool) {
+
+	log.Debug("Handle controlMapPause state")
+
+	// Schedule a timer for the next update map event to
+	// fetch from the server
+	nextMapRefresh, err := controller.GetControlMapPool().
+		NextControlMapHalfTime(c.wrappedState.Update().ID)
+
+	if errors.Is(err, NoUpdateMapsErr) {
+		log.Error("No control maps no longer present, continuing")
+		nextMapRefresh = time.Now().Add(30 * time.Second)
```

Comment:
Hmm, I thought about it, and I sort of fell down upon this, that the safest is just to continue the regular execution.
However, I see what you mean, and in that case, maybe going back to `controlMapState` is the best option? Since then all exists from the control "loop" are at the same point (less magic) ? :)

Reply:
Yeah, I don't have a strong feeling about it, but it does seem preferable to me too.